File: Assets/Scripts/Core/SpecialCardEffectResult.cs
namespace Bunker.Core
{
    // Describes what actually happened after an effect was applied.
    // Sent back to the caller (and, later, over the network) so the UI knows
    // what to show — e.g. a revealed trait should only be shown to the caster,
    // not broadcast to everyone.
    public class SpecialCardEffectResult
    {
        public bool Success;
        public string FailReason;

        // Set only for effects that reveal private info to the caster
        // (e.g. RevealHiddenTrait). Null for effects with no private payload.
        public CharacterTrait RevealedTrait;

        public static SpecialCardEffectResult Fail(string reason) =>
            new() { Success = false, FailReason = reason };

        public static SpecialCardEffectResult Ok(CharacterTrait revealedTrait = null) =>
            new() { Success = true, RevealedTrait = revealedTrait };
    }
}

File: bunker-server/src/game/effects/SpecialCardEffectResult.ts
import { CharacterTrait } from "../CharacterTrait";

// Equivalent of the C# SpecialCardEffectResult (section 1.5).
export interface SpecialCardEffectResult {
  success: boolean;
  failReason?: string;
  revealedTrait?: CharacterTrait; // set only for effects with a private payload (RevealHiddenTrait)
}

export function fail(reason: string): SpecialCardEffectResult {
  return { success: false, failReason: reason };
}

export function ok(revealedTrait?: CharacterTrait): SpecialCardEffectResult {
  return { success: true, revealedTrait };
}
